extensions/data-plane/data-plane-aws-s3/src/main/java/org/eclipse/edc/connector/dataplane/aws/s3/S3DataSource.java

```diff
@@ -14,28 +14,58 @@
 
 package org.eclipse.edc.connector.dataplane.aws.s3;
 
+import org.eclipse.edc.connector.dataplane.aws.s3.exception.S3ObjectNotFoundException;
 import org.eclipse.edc.connector.dataplane.spi.pipeline.DataSource;
+import org.eclipse.edc.connector.dataplane.spi.pipeline.StreamFailure;
 import org.eclipse.edc.connector.dataplane.spi.pipeline.StreamResult;
 import software.amazon.awssdk.services.s3.S3Client;
 import software.amazon.awssdk.services.s3.model.GetObjectRequest;
 import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.S3Object;
 
 import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.stream.Stream;
 
+import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamFailure.Reason.GENERAL_ERROR;
+import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamResult.failure;
 import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamResult.success;
 
 class S3DataSource implements DataSource {
 
     private String bucketName;
     private String keyName;
+    private String keyPrefix;
     private S3Client client;
 
     private S3DataSource() {
     }
 
     @Override
     public StreamResult<Stream<Part>> openPartStream() {
+
+        if (keyPrefix != null) {
+            try {
+
+                var s3Objects = this.fetchPrefixedS3Objects();
+
+                if (s3Objects.isEmpty()) {
+                    throw new S3ObjectNotFoundException("Object not found");
```

Comment:
using exceptions to describe the application flow is overkill, here it could already return the `StreamResult.failure` and the wrapping `try/catch` block could be removed as well